File: snippets/java/src/main/java/develop/Awakeables.java
package develop;

import dev.restate.sdk.Awakeable;
import dev.restate.sdk.Restate;

public class Awakeables {

  public void awakeables() {
    String name = "hello";
    // <start_here>
    // Create awakeable and get unique ID
    Awakeable<String> awakeable = Restate.awakeable(String.class);
    String awakeableId = awakeable.id();

    // Send ID to external system (email, queue, webhook, etc.)
    Restate.run("request-human-review", () -> requestHumanReview(name, awakeableId));

    // Handler suspends here until external completion
    String review = awakeable.await();
    // <end_here>

    // <start_resolve>
    // Complete with success data
    Restate.awakeableHandle(awakeableId).resolve(String.class, "Looks good!");
    // <end_resolve>

    // <start_reject>
    // Complete with error
    Restate.awakeableHandle(awakeableId).reject("This cannot be reviewed.");
    // <end_reject>
  }

  public void requestHumanReview(String name, String awakeableId) {}
}

File: snippets/kotlin/src/main/kotlin/develop/Awakeables.kt
package develop

import dev.restate.sdk.kotlin.*

class Awakeables {
  suspend fun awakeables() {
    // <start_here>
    // Create awakeable and get unique ID
    val awakeable = awakeable<String>()
    val awakeableId = awakeable.id

    // Send ID to external system (email, queue, webhook, etc.)
    runBlock { requestHumanReview(awakeableId) }

    // Handler suspends here until external completion
    val review = awakeable.await()
    // <end_here>

    // <start_resolve>
    // Complete with success data
    awakeableHandle(awakeableId).resolve("Looks good!")
    // <end_resolve>

    // <start_reject>
    // Complete with error
    awakeableHandle(awakeableId).reject("This cannot be reviewed.")
    // <end_reject>
  }

  private fun requestHumanReview(awakeableId: String): String {
    return "hello"
  }
}
